src/app/shopping-cart.service.ts

```diff
@@ -0,0 +1,66 @@
+import { Injectable } from '@angular/core';
+import { AngularFireDatabase } from '@angular/fire/database';
+import { Product } from './models/product';
+import { take } from 'rxjs/operators';
+import { Observable } from 'rxjs';
+import { ShoppingCartItem } from './models/shoppingCartItem';
+import { map } from 'rxjs/operators';
+import { ShoppingCart } from './models/shoppingCart';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class ShoppingCartService {
+  itemObservable;
+  constructor(private database: AngularFireDatabase) {}
+
+  private create() {
+    return this.database.list('/shopping-carts').push({
+      data: new Date().getTime()
+    });
+  }
+
+  getCart(): Observable<ShoppingCart> {
+    let cartId = this.getOrCreateCartId();
+    return this.database
+      .object('/shopping-carts/' + cartId)
+      .valueChanges() as Observable<ShoppingCart>;
+  }
+
+  private getOrCreateCartId() {
+    let cartId = localStorage.getItem('cartId');
+    if (cartId) return cartId;
+    let result = this.create();
+    localStorage.setItem('cartId', result.key);
+    return result.key;
+  }
+
+  private getItem(cartId: string, productId: string) {
+    return this.database.object(
+      '/shopping-carts/' + cartId + '/items/' + productId
+    );
+  }
+
+  addToCart(product: Product) {
+    this.updateItemQuantity(product, 1);
+  }
+
+  removeFromCart(product: Product) {
+    this.updateItemQuantity(product, -1);
+  }
+
+  private async updateItemQuantity(product: Product, change: number) {
+    let cartId = this.getOrCreateCartId();
+    this.itemObservable = this.getItem(cartId, product.id);
+    this.itemObservable
+      .snapshotChanges()
+      .pipe(take(1))
+      .subscribe((item: any) => {
+        this.itemObservable.update({
```

Comment:
Do you need to use both take() and subscribe here?  I haven't gotten to this part yet but it was my understanding that they both can be used to subscribe to Observables.

Reply:
I think I need it because when I took it off just now clicking the plus button ended up with 900 breads in my cart. lol